File: apps/go-idp/internal/queries/verify_token.go
package queries

import (
	"context"

	"go-idp/internal/auth"
)

type VerifyTokenInput struct {
	AccessToken string
}

type VerifyTokenOutput struct {
	Valid  bool    `json:"valid"`
	UserID *string `json:"user_id"`
	Exp    *int64  `json:"exp"`
}

type VerifyToken struct {
	signer *auth.JWTSigner
}

func NewVerifyToken(signer *auth.JWTSigner) *VerifyToken {
	return &VerifyToken{signer: signer}
}

func (q *VerifyToken) Handle(_ context.Context, in VerifyTokenInput) *VerifyTokenOutput {
	claims, err := q.signer.Verify(in.AccessToken)
	if err != nil || claims == nil {
		return &VerifyTokenOutput{Valid: false}
	}
	sub := claims.Sub
	exp := claims.Exp
	return &VerifyTokenOutput{Valid: true, UserID: &sub, Exp: &exp}
}

File: apps/hono-idp/src/queries/verifyToken.ts
import { verifyAccessToken } from '../auth/jwt';

export type VerifyTokenInput = { access_token: string };
export type VerifyTokenOutput =
  | { valid: true; user_id: string; exp: number }
  | { valid: false; user_id: null; exp: null };

export function verifyTokenQuery() {
  return async (input: VerifyTokenInput): Promise<VerifyTokenOutput> => {
    const claims = await verifyAccessToken(input.access_token);
    if (!claims) return { valid: false, user_id: null, exp: null };
    return { valid: true, user_id: claims.sub, exp: claims.exp };
  };
}
